core/src/main/java/org/mule/runtime/core/internal/context/thread/notification/ThreadLoggingExecutorServiceDecorator.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.context.thread.notification;
+
+import org.mule.runtime.core.api.event.CoreEvent;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Decorates a {@link ExecutorService} to add logging. This is useful in cases like subscribeOn
+ * that impose a Thread switch at the construction of the chain
+ *
+ * @since 4.2
+ */
+public class ThreadLoggingExecutorServiceDecorator implements ExecutorService {
+
+  private ExecutorService delegate;
+  private Optional<ThreadNotificationLogger> threadNotificationLogger;
+  private CoreEvent event;
```

Comment:
are you using anything other that the id from the event? If not, keep just the id to avoid keeping references to an Event